### ISSUE TITLE
HDF5 format

### ISSUE THREAD

**Issue by Jullija** [psrenergy/QUBOTools.jl#47]
I am working on adding a new format to QUBOTools.jl - HDF5. 
In mentioned format in dataset I would like to have two-dimensional matrix. Model allows only dictionaries for linear and quadratic terms. 
Yet, rewriting an array to dictionary takes plenty of time. I would like to skip this step and write it as matrix. 
I was thinking about adding reading and writing terms for matrix to every other format, but I don't think it's the best way. Do you have any idea how to do it better?

**Comment by bernalde** [comment]
Hello, we are working on a private branch on this idea, and it might make sense for us to compare notes. I'll bring @pedromxavier to the table and we can have a more involved conversation

**Comment by pedromxavier** [comment]
Hi @Jullija, we're currently working on the [px/refactor-v1](https://github.com/psrenergy/QUBOTools.jl/tree/px/refactor-v1) branch, with its open PR (#44). As you can see there, there are many design choices under discussion, so I would say it's the best moment to raise this kind of question.

Something that is changing is that the Model now stores (sparse) arrays instead of dictionaries. Some operations are faster and others aren't, but I believe it will behave better in the general case where the main task is to move data around. Another option is to make this in-memory storage more flexible, allowing for dense matrices depending on the input, for example.

I've prepared a HDF5-based format that stores these sparse matrices, using the `SparseArrays` API. I'm really interested in knowing more about your format proposal and your use cases before proceeding.

I'm adding this issue to the PR checklist. Thank you!

**Comment by Jullija** [comment]
Hi, thank you for your answer!

When I tried to add HDF5 format, I used dictionaries so my code will not help at all. 

I am a part of quantumz.io and we are working on Multi-purpose Algorithmic Generator for Instances of Combinatorial Optimization. We would like to use QUBOTools for our task.

We need to store files in such a format:
Dense:
<img width="602" alt="245054723-f1629964-45f8-4b84-8a43-4897fe689df4" src="https://github.com/psrenergy/QUBOTools.jl/assets/97158770/26b88ecd-ac7d-4876-8d5c-d08a0ad9575b">
Sparse:
<img width="621" alt="245054764-8b378b83-1435-4289-ba1b-9638805d5c58" src="https://github.com/psrenergy/QUBOTools.jl/assets/97158770/b9b8071a-f73f-48cf-ad1d-bb28c8452b91">

**Comment by pedromxavier** [comment]
Our initial proposal for the HDF5-based format (`QUBOTools.QUBin`, extension `.qb`) is the following:

```julia
julia> using HDF5

julia> fp = HDF5.h5open("test/data/02/bool.qb", "r")
🗂️ HDF5.File: (read-only) test/data/02/bool.qb
├─ 📂 model
│  ├─ 📂 form
│  │  ├─ 🔢 dimension
│  │  ├─ 🔢 domain
│  │  ├─ 📂 linear
│  │  │  ├─ 🔢 i
│  │  │  └─ 🔢 v
│  │  ├─ 🔢 offset
│  │  ├─ 📂 quadratic
│  │  │  ├─ 🔢 i
│  │  │  ├─ 🔢 j
│  │  │  └─ 🔢 v
│  │  ├─ 🔢 scale
│  │  └─ 🔢 sense
│  ├─ 🔢 metadata
│  └─ 🔢 variables
└─ 📂 solution
   ├─ 📂 data
   │  ├─ 🔢 reads
   │  ├─ 🔢 state
   │  └─ 🔢 value
   ├─ 🔢 domain
   ├─ 🔢 metadata
   └─ 🔢 sense
```

Currently, it is based on the sparse matrix layout, but we can think about making this more flexible too.

Thanks to you input and some previous thoughts, I've just implemented in #44 a mechanism for allowing our `Model` to represent data using dictionaries, dense and sparse arrays via parametric types. This is the first step towards being able to choose which underlying data structure to employ depending on the application. It still requires testing and a few checks on type stability across those operation modes.

We're happy that quantumz.io is interested in using the tools. We are open for discussing these and other development tasks, so feel free to get in touch. Here is [my contact page](https://pedromxavier.github.io/contact).
